picocli-shell-jline3/src/test/java/picocli/shell/jline3/example/Example.java

```diff
@@ -172,7 +152,7 @@ public static void main(String[] args) {
                         .variable(LineReader.LIST_MAX, 50)   // max tab completion candidates
                         .build();
                 builtins.setLineReader(reader);
-                commands.setReader(reader);
+                picocliCommands.includeClearScreenCommand(reader);
```

Comment:
If I understand correctly, the above line creates a `ClearCommand` with the specified `LineReader`, and adds it to the command hierarchy. 

I would prefer if we could add the command declaratively with `@Command(subcommands = ClearCommand.class)`. To do that, we need some way to pass the `LineReader` to the `ClearCommand` (without relying on `@ParentCommand`).

One idea is to do it with a custom factory like this:

```java
LineReader lineReader = ... // create line reader
IFactory factory = new IFactory() {
    public <K> K create(Class<K> clazz) throws Exception {
        if (ClearCommand.class == clazz) { return new ClearCommand(lineReader); }
        return CommandLine.defaultFactory().create(clazz); 
    }
}
CommandLine cmd = new CommandLine(commands, factory);
...
```

Thoughts?

Reply:
@remkop - this is a better way to use the picocli API, of course. However there is a circular dependency in how these are created:
line reader -> system registry completer -> picocli commands -> line reader
In order to break this cyclic relationship, we need to construct the clear command later on, and pass the line reader in. Can you think of another way to break the cycle?

Reply:
Does this work?

```java
// set up picocli commands
class MyFactory implements CommandLine.IFactory {
    LineReader reader;
    public <K> K create(Class<K> clazz) throws Exception {
        if (ClearCommand.class == clazz) { return new ClearCommand(reader); }
        return CommandLine.defaultFactory().create(clazz);
    }
}
MyFactory myFactory = new MyFactory();
CliCommands commands = new CliCommands();
CommandLine cmd = new CommandLine(commands, myFactory);
PicocliCommands picocliCommands = new PicocliCommands(Example::workDir, cmd);

Parser parser = new DefaultParser();
try (Terminal terminal = TerminalBuilder.builder().build()) {
    SystemRegistry systemRegistry = new SystemRegistryImpl(parser, terminal, Example::workDir, null);
    systemRegistry.setCommandRegistries(builtins, picocliCommands);

    LineReader reader = LineReaderBuilder.builder()
            .terminal(terminal)
            .completer(systemRegistry.completer())
            .parser(parser)
            .variable(LineReader.LIST_MAX, 50)  
            .build();
    myFactory.reader = reader; // now pass the reader to the factory
```

Reply:
@remkop - I have moved the `ClearCommand` class to `PicocliCommands`, so the users of picocli do not have to deal with it. The idea is that a user can turn on the clear command with one method call on `PicocliCommands` (and similarly turn on picocli help with one method call).

Reply:
ok, perhaps the factory should also live in PicocliCommands, so it is more in one place.

Reply:
Sure, but what if someone uses their own factory? (I use a custom factory to maintain state in the jline3 shell. The factory injects the same state object instance into each command.) If we are going this route, picocli would need to allow a factory chain. Maybe that is a feature request for picocli itself?

Reply:
Okay, so it should be something like:

```java
class MyFactory implements CommandLine.IFactory {
    LineReader reader;
    CommandLine.IFactory fallbackFactory;

    public MyFactory() {
        this(CommandLine.defaultFactory());
    }

    public MyFactory(CommandLine.IFactory fallbackFactory) {
        this.fallbackFactory = Objects.requireNonNull(fallbackFactory);
    }

    public <K> K create(Class<K> clazz) throws Exception {
        if (ClearCommand.class == clazz) { return new ClearCommand(reader); }
        return fallbackFactory.create(clazz);
    }
}
```

Reply:
Thanks.

I will need to modify the `PicocliCommands` constructor (maybe add a boolean argument) , and add in the clear command to the passed in command-line object. If I am going to modify the passed in `CommandLine` object anyway, I can wrap in the factory with `MyFactory` internally in the `PicocliCommands` constructor. In order to do this, I would need a getter to get the original factory on the original `CommandLine`. (I could not find this getter on `CommandLine`.)

This is all probably making it more convoluted than my original solution. What do you think?

Reply:
PS: Please consider making the factory follow a chain of responsibility design pattern in a future release. Would you like me to create an issue for this?

Reply:
> I will need to modify the PicocliCommands constructor (maybe add a boolean argument) and add in the clear command to the passed in command-line object

Why? I was thinking of just moving the factory to the `PicocliCommands` class. Something like this:

```java
PicocliCommands.Factory myFactory = new PicocliCommands.Factory();
CliCommands commands = new CliCommands();
CommandLine cmd = new CommandLine(commands, myFactory);
PicocliCommands picocliCommands = new PicocliCommands(Example::workDir, cmd);

Parser parser = new DefaultParser();
try (Terminal terminal = TerminalBuilder.builder().build()) {
    SystemRegistry systemRegistry = new SystemRegistryImpl(parser, terminal, Example::workDir, null);
    systemRegistry.setCommandRegistries(builtins, picocliCommands);

    LineReader reader = LineReaderBuilder.builder()
            .terminal(terminal)
            .completer(systemRegistry.completer())
            .parser(parser)
            .variable(LineReader.LIST_MAX, 50)  
            .build();
    myFactory.setReader(reader); // now pass the reader to the factory
```


> PS: Please consider making the factory follow a chain of responsibility design pattern in a future release. Would you like me to create an issue for this?

Factory implementors can easily implement their own chain of responsibility if they want. One simple way to do this is with a constructor that accepts another factory as in the code snippet I posted in my previous comment. I do not see what value a chain of responsibility would add, but perhaps I am missing something.

Reply:
PS. I created https://github.com/remkop/picocli/issues/1267 for creating an accessor for the factory. We don't need it for our use case here, but it is nice to have.

Reply:
@sualeh I saw your message on the picocli Google group, but I think there may be a misunderstanding. 
I think we are almost there.

I did not understand your comment:

> I will need to modify the `PicocliCommands` constructor (maybe add a boolean argument) , and add in the clear command to the passed in command-line object. If I am going to modify the passed in `CommandLine` object anyway, I can wrap in the factory with `MyFactory` internally in the `PicocliCommands` constructor. In order to do this, I would need a getter to get the original factory on the original `CommandLine`. (I could not find this getter on `CommandLine`.)

Perhaps I misunderstand, but I think we can leave the `PicocliCommands` constructor as is, or am I missing something?

About the factory, and combining it with a custom factory, I imagine applications would use it like this:

```java
MyCustomFactory customFactory = createCustomFactory(); // your application custom factory
PicocliCommands.Factory factory = new PicocliCommands.Factory(customFactory); // chain the factories
CliCommands commands = new CliCommands();
CommandLine cmd = new CommandLine(commands, factory);
...
factory.setReader(reader); // set LineReader after it is created
```

This way, we no longer need the `PicocliCommands.includeClearScreenCommand` method, and applications can use the new ClearScreen command declaratively with `@Command(subcommands = { PicocliCommands.ClearScreen.class, ...})`.

This looks nice to me, but what do you think?